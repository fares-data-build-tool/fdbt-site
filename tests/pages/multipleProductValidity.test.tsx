/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { shallow } from 'enzyme';
import React from 'react';
import MultiProductValidity, { getServerSideProps } from '../../src/pages/multipleProductValidity';
import { getMockContext } from '../testData/mockData';

describe('pages', () => {
    describe('multipleProductValidity', () => {
        const wrapper = shallow(
            <MultiProductValidity
                operator="Infinity Line"
                numberOfProducts="2"
                multipleProducts={[
                    {
                        productName: 'super ticket',
                        productPrice: '3.50',
                        productDuration: '3',
                    },
                    {
                        productName: 'best ticket',
                        productPrice: '30.90',
                        productDuration: '30',
                    },
                ]}
                errors={[]}
            />,
        );

        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders 2 radio buttons per number of products', () => {
            expect(wrapper.find('.govuk-radios__item')).toHaveLength(4);
        });
    });

    describe('server side props', () => {
        it('should return number of products to display, name of operator and products if they are set in the cookie', () => {
            const ctx = getMockContext();
            const result = getServerSideProps(ctx);

            expect(result.props.numberOfProducts).toBe('2');
            expect(result.props.operator).toBe('test');
            expect(result.props.multipleProducts[0].productName).toBe('Best Product');
            expect(result.props.multipleProducts[0].productPrice).toBe('2');
            expect(result.props.multipleProducts[0].productDuration).toBe('3');
            expect(result.props.multipleProducts.length).toBe(3);
        });

        it('should throw an error if the necessary cookies to render the page are not present', () => {
            const ctx = getMockContext({
                operator: null,
                numberOfProductsInput: null,
                multipleProduct: null,
            });
            expect(() => getServerSideProps(ctx)).toThrow(
                'Necessary cookies not found to display the multiple product vaidity page',
            );
        });

        it('returns errors in the props if there are validity errors on the product object', () => {
            const ctx = getMockContext({
                multipleProducts: [
                    {
                        productName: 'Best Product',
                        productNameId: 'multipleProductNameInput0',
                        productPrice: '2',
                        productPriceId: 'multipleProductPriceInput0',
                        productDuration: '3',
                        productDurationId: 'multipleProductDurationInput0',
                        productValidity: {
                            validity: '',
                            error: 'Select one of the two validity options',
                        }
                    },
                    {
                        productName: 'Super Product',
                        productNameId: 'multipleProductNameInput1',
                        productPrice: '3',
                        productPriceId: 'multipleProductPriceInput1',
                        productDuration: '4',
                        productDurationId: 'multipleProductDurationInput1',
                        productValidity: {
                            validity: '24hrs',
                            error: '',
                        }
                    },
                ],
            });

            const result = getServerSideProps(ctx);

            expect(result.props.multipleProducts.length).toBe(2);
            expect(result.props.errors.length).toBe(1);
            expect(result.props.errors[0].errorMessage).toBe('Select one of the two validity options');
            expect(result.props.errors[0].id).toBe('multiple-product-validity-radios-error');
        });
    });
});
