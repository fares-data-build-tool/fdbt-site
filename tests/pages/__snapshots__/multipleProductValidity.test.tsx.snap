// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages multipleProductValidity should render correctly 1`] = `
<Component
  description="Multiple Product Validity selection page of the Create Fares Data Service"
  errors={Array []}
  title="Multiple Product Validity - Create Fares Data Service"
>
  <CsrfForm
    action="/api/multipleProductValidity"
    className="multiple-product-validity-page"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <div
      className="govuk-form-group "
    >
      <h1
        className="govuk-heading-l"
        id="multiple-product-validity-page-heading"
      >
        When does the product expire?
      </h1>
      <span
        className="govuk-hint"
        id="operator-products-hint"
      >
        Infinity Line
         - 
        2
         products - 
        Adult
      </span>
      <span
        className="govuk-hint"
        id="multiple-product-validity-page-hint"
      >
        We need to know the time that this product would be valid until
      </span>
      <span
        className="govuk-hint"
        id="24hr-validity-type-hint"
      >
        24hr means a ticket purchased at 3pm will be valid until 3pm on its day of expiry
      </span>
      <span
        className="govuk-hint"
        id="calendar-validity-type-hint"
      >
        Calendar day means a ticket purchased at 3pm would be valid until midnight on its day of expiry
      </span>
      <FormElementWrapper
        errorClass="govuk-radios--error"
        errorId="twenty-four-hours-row-0"
        errors={Array []}
      >
        <table
          className="govuk-table multiple-product-validity-table"
        >
          <thead
            className="govuk-table__head"
          >
            <tr
              className="govuk-table__row"
            >
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Name
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Price
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Duration
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Choose Validity
              </th>
            </tr>
          </thead>
          <tbody
            className="govuk-table__body"
          >
            <tr
              className="govuk-table__row"
              key="multiple-product-name-1"
            >
              <td
                className="govuk-table__cell"
              >
                Weekly Ticket
              </td>
              <td
                className="govuk-table__cell"
              >
                £
                50
              </td>
              <td
                className="govuk-table__cell"
              >
                5 undefineds
              </td>
              <td
                className="govuk-table__cell"
              >
                <select
                  aria-labelledby="stop-name-header stop-0 naptan-code-header naptan-0"
                  className="govuk-select farestage-select"
                  id="option-0"
                  name="option-0"
                >
                  <option
                    disabled={true}
                    selected={true}
                    value=""
                  >
                    Select validity
                  </option>
                  <option
                    value="24hr"
                  >
                    24hr
                  </option>
                  <option
                    value="endOfCalendarDay"
                  >
                    Calendar day
                  </option>
                  <option
                    value="serviceDay"
                  >
                    Service day
                  </option>
                </select>
              </td>
            </tr>
            <tr
              className="govuk-table__row"
              key="multiple-product-name-2"
            >
              <td
                className="govuk-table__cell"
              >
                Day Ticket
              </td>
              <td
                className="govuk-table__cell"
              >
                £
                2.50
              </td>
              <td
                className="govuk-table__cell"
              >
                1 undefined
              </td>
              <td
                className="govuk-table__cell"
              >
                <select
                  aria-labelledby="stop-name-header stop-1 naptan-code-header naptan-1"
                  className="govuk-select farestage-select"
                  id="option-1"
                  name="option-1"
                >
                  <option
                    disabled={true}
                    selected={true}
                    value=""
                  >
                    Select validity
                  </option>
                  <option
                    value="24hr"
                  >
                    24hr
                  </option>
                  <option
                    value="endOfCalendarDay"
                  >
                    Calendar day
                  </option>
                  <option
                    value="serviceDay"
                  >
                    Service day
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </FormElementWrapper>
    </div>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;

exports[`pages multipleProductValidity should render error messaging when errors are passed 1`] = `
<Component
  description="Multiple Product Validity selection page of the Create Fares Data Service"
  errors={
    Array [
      Object {
        "errorMessage": "Select one of the two validity options",
        "id": "twenty-four-hours-row-0",
      },
    ]
  }
  title="Multiple Product Validity - Create Fares Data Service"
>
  <CsrfForm
    action="/api/multipleProductValidity"
    className="multiple-product-validity-page"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={
        Array [
          Object {
            "errorMessage": "Select one of the two validity options",
            "id": "twenty-four-hours-row-0",
          },
        ]
      }
    />
    <div
      className="govuk-form-group govuk-form-group--error"
    >
      <h1
        className="govuk-heading-l"
        id="multiple-product-validity-page-heading"
      >
        When does the product expire?
      </h1>
      <span
        className="govuk-hint"
        id="operator-products-hint"
      >
        Infinity Line
         - 
        2
         products - 
        Adult
      </span>
      <span
        className="govuk-hint"
        id="multiple-product-validity-page-hint"
      >
        We need to know the time that this product would be valid until
      </span>
      <span
        className="govuk-hint"
        id="24hr-validity-type-hint"
      >
        24hr means a ticket purchased at 3pm will be valid until 3pm on its day of expiry
      </span>
      <span
        className="govuk-hint"
        id="calendar-validity-type-hint"
      >
        Calendar day means a ticket purchased at 3pm would be valid until midnight on its day of expiry
      </span>
      <FormElementWrapper
        errorClass="govuk-radios--error"
        errorId="twenty-four-hours-row-0"
        errors={
          Array [
            Object {
              "errorMessage": "Select one of the two validity options",
              "id": "twenty-four-hours-row-0",
            },
          ]
        }
      >
        <table
          className="govuk-table multiple-product-validity-table"
        >
          <thead
            className="govuk-table__head"
          >
            <tr
              className="govuk-table__row"
            >
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Name
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Price
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Product Duration
              </th>
              <th
                className="govuk-table__header govuk-!-width-one-quarter"
                scope="col"
              >
                Choose Validity
              </th>
            </tr>
          </thead>
          <tbody
            className="govuk-table__body"
          >
            <tr
              className="govuk-table__row"
              key="multiple-product-name-1"
            >
              <td
                className="govuk-table__cell"
              >
                Weekly Ticket
              </td>
              <td
                className="govuk-table__cell"
              >
                £
                50
              </td>
              <td
                className="govuk-table__cell"
              >
                5 undefineds
              </td>
              <td
                className="govuk-table__cell"
              >
                <select
                  aria-labelledby="stop-name-header stop-0 naptan-code-header naptan-0"
                  className="govuk-select farestage-select"
                  id="option-0"
                  name="option-0"
                >
                  <option
                    disabled={true}
                    selected={true}
                    value=""
                  >
                    Select validity
                  </option>
                  <option
                    value="24hr"
                  >
                    24hr
                  </option>
                  <option
                    value="endOfCalendarDay"
                  >
                    Calendar day
                  </option>
                  <option
                    value="serviceDay"
                  >
                    Service day
                  </option>
                </select>
              </td>
            </tr>
            <tr
              className="govuk-table__row"
              key="multiple-product-name-2"
            >
              <td
                className="govuk-table__cell"
              >
                Day Ticket
              </td>
              <td
                className="govuk-table__cell"
              >
                £
                2.50
              </td>
              <td
                className="govuk-table__cell"
              >
                1 undefined
              </td>
              <td
                className="govuk-table__cell"
              >
                <select
                  aria-labelledby="stop-name-header stop-1 naptan-code-header naptan-1"
                  className="govuk-select farestage-select"
                  id="option-1"
                  name="option-1"
                >
                  <option
                    disabled={true}
                    selected={true}
                    value=""
                  >
                    Select validity
                  </option>
                  <option
                    value="24hr"
                  >
                    24hr
                  </option>
                  <option
                    value="endOfCalendarDay"
                  >
                    Calendar day
                  </option>
                  <option
                    value="serviceDay"
                  >
                    Service day
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </FormElementWrapper>
    </div>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;
